File: Chapter 3/Section 3.3/range.cpp
/*
ID: hsfncd31
TASK: range
LANG: C++                 
*/
#include <fstream>

int main() {
    std::fstream in("range.in", std::fstream::in), out("range.out", std::fstream::out);

    static const int MAX_N = 250;
    int n;
    std::string s;

    bool field[2][MAX_N][MAX_N];
    bool (&current)[MAX_N][MAX_N] = field[0], (&last)[MAX_N][MAX_N] = field[1];

    in >> n;
    for (int row = 0; row < n; ++row) {
        in >> s;
        for (int column = 0; column < n; ++column)
            current[row][column] = (s[column] == '1');
    }

    for (int size = 2; ; ++size) {
        int num_larger = 0;
        std::swap(current, last);
        for (int row = 0; row < n - size + 1; ++row)
            for (int column = 0; column < n - size + 1; ++column) {
                bool larger = (last[row][column] && last[row][column + 1] && last[row + 1][column] && last[row + 1][column + 1]);
                current[row][column] = larger;
                num_larger += larger;
            }
        if (num_larger)
            out << size << ' ' << num_larger << '\n';
        else
            break;
    }

    in.close();
    out.close();
}

File: Chapter 3/Section 3.3/range.py
"""
ID: hsfncd31
LANG: PYTHON3
TASK: range
TLE at Case#7
"""
import os
import typing
import itertools


def main():
    base_filename = 'test' if os.name == 'nt' else 'range'
    with open('{}.in'.format(base_filename), 'r') as infile,\
            open('{}.out'.format(base_filename), 'w') as outfile:  # type: typing.IO[str]
        def get_line() -> str:
            return infile.readline().rstrip('\n')

        def out_print(*args, **kwargs) -> None:
            kwargs['file'] = outfile
            print(*args, **kwargs)

        # Mark the upper-left corner of each square.
        # Starting from size 2, four marks in a 2x2 square means a larger square.
        n = int(get_line())
        field = [[bool(c == '1') for c in get_line()] for _ in range(n)]
        current = field.copy()
        last = [[False] * n for _ in range(n)]
        for size in itertools.count(2):
            current, last = last, current
            num_larger = 0
            for row in range(n - size + 1):
                for column in range(n - size + 1):
                    larger_square = bool(last[row][column] and last[row][column + 1]
                                         and last[row + 1][column] and last[row + 1][column + 1])
                    current[row][column] = larger_square
                    num_larger += 1 if larger_square else 0
            if num_larger:
                out_print(size, num_larger)
            else:
                break


if __name__ == '__main__':
    main()
